Week 9/data_cleansing_and_transformation.ipynb

```diff
@@ -0,0 +1,311 @@
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(8), np.int64(10)}\n",
+      "Index(['pdays_1', 'pdays_2', 'pdays_3', 'pdays_4', 'pdays_5', 'pdays_6',\n",
+      "       'pdays_7', 'pdays_8', 'pdays_9', 'pdays_10', 'pdays_11', 'previous',\n",
+      "       'nr.employed'],\n",
+      "      dtype='object')\n"
+     ]
+    }
+   ],
+   "source": [
+    "import matplotlib.pyplot as plt \n",
+    "\n",
+    "numerical_df = df[numeric_columns]\n",
+    "\n",
+    "new_categorical_vars = pd.DataFrame()\n",
+    "\n",
+    "# print(len(numerical_df[\"pdays\"][numerical_df[\"pdays\"]!=999]))\n",
+    "# print(len(numerical_df[\"pdays\"][numerical_df[\"pdays\"]==999]))\n",
+    "# print(len(numerical_df[\"pdays\"]))\n",
+    "\n",
+    "\n",
+    "# The pdays feature is extremely abnormal due to many data points with the value of 999\n",
+    "# it is binned in a special way to avoid many bins with zero frequencies\n",
+    "# first the data is filtered for values that are not 999\n",
+    "# Then normal binning commences\n",
+    "# then the non filtered series is used to add the values of 999 with a corresponding bin value of 11\n",
+    "# this feature should be transformed into a onehot-encoded categorical variable since \n",
+    "# the values between what the bins represent is non-linear since values with 999 is in the last bin.\n",
+    "\n",
+    "\n",
+    "filtered_pday = numerical_df[\"pdays\"][numerical_df[\"pdays\"]!=999]\n",
+    "bins = np.linspace(min(filtered_pday), max(filtered_pday), 10) \n",
+    "binned = list(np.digitize(filtered_pday, bins))\n",
+    "old_nof_values = len(set(binned))\n",
+    "i =0\n",
+    "for item in list(numerical_df[\"pdays\"]):\n",
+    "    if item == 999:\n",
+    "        binned.insert(i, np.int64(old_nof_values+1))\n",
+    "i+=1\n",
+    "plt.hist(binned, bins=len(bins)+1, edgecolor='black') \n",
+    "plt.show()\n",
+    "new_categorical_vars = pd.get_dummies(pd.Categorical(binned), prefix='pdays')\n",
+    "\n",
+    "\n",
+    "# Convert previous to ordinal categorical variable due to extreme skewness\n",
+    "print(numerical_df[\"previous\"].unique())\n",
+    "new_categorical_vars[\"previous\"] = pd.Categorical(numerical_df[\"previous\"], ordered=True)\n",
+    "print(new_categorical_vars[\"previous\"].unique())\n",
+    "print(new_categorical_vars.columns)\n",
+    "\n",
+    "\n",
+    "# convert nr.employed to ordinal categorical variable with binning\n",
+    "# This is due to skewness that could not be solved in log and cube transformations\n",
+    "# Note: When outliers are removed\n",
+    "bins = np.linspace(min(numerical_df[\"nr.employed\"]), max(numerical_df[\"nr.employed\"]), 10) \n",
+    "binned = list(np.digitize(numerical_df[\"nr.employed\"], bins))\n",
+    "plt.hist(binned, bins=len(bins), edgecolor='black') \n",
```

Comment:
The histogram of `pdays` after binning is a good step for visualizing data distribution. For better understanding, it might be useful to add labels to the x-axis and y-axis.